### ISSUE TITLE
Strange `eltype` bahavior

### ISSUE THREAD

**Issue by cscherrer** [JuliaStats/Distributions.jl#1402]
This seems problematic:
```julia
julia> using Distributions

julia> eltype(Normal())
Float64

julia> eltype(MvNormal(ones(3)))
Float64
```

I found this very surprising, as I'd guess many others have/will. In general, it seems reasonable to expect a law that for any `d::Distribution`,
```julia
typeof(rand(d)) <: eltype(d)
```

My biggest concern here is that I'd like to implement `eltype` in MeasureTheory.jl, and I don't know if people might expect it to behave as it does in Distributions. 

Could there be a path to changing this?

**Comment by cscherrer** [comment]
Thanks for the link! Looks like this has been quite a saga. This design decision for `eltype` to refer to the parameters is very confusing, and seems inconsistent with the rest of Julia. I think it may be best for MeasureTheory to ignore this quirk and just focus of having its `eltype` consistent with Base.

**Comment by devmotion** [comment]
> which should then match the type that comes out of rand.

The main issue here is that it is not always possible to compute the type of `rand(d)` in a nice way, e.g., if sampling is based on a complicated push forward operation or you want to use different "elementary" random samples, e.g. `rand()`, `rand(Float32)`, `rand(Float16)`, or `rand(BigFloat)` for different precision.

**Comment by mschauer** [comment]
Yeah, you know there is a parallel infrastructure where the type of `rand(d)` is derived from `ValueSupport` and `VariateForm` *in most cases* https://github.com/JuliaStats/Distributions.jl/blob/625b72237a342c8d3bf60ec05541f8cb4a78faff/src/common.jl

**Comment by devmotion** [comment]
Sure but unfortunately it is only a heuristic and requires e.g. `float` in https://github.com/JuliaStats/Distributions.jl/blob/27fb31ce6098149fa051613483edfafd616de310/src/univariates.jl#L140.

**Comment by devmotion** [comment]
> It wouldn't return an element type, but the full type of a single variate, so Float64, or Int, or Vector{Float32}, etc.

I think the only reliable way that always works is to call `typeof(rand(d))` :shrug: I guess for many simple cases a heuristic such as `float(eltype(d))` for continuous univariate distributions will work but it's not guaranteed to be correct.

**Comment by oschulz** [comment]
> I think the only reliable way that always works ... I guess for many simple cases a heuristic

Well, however implements `rand` should be able to predict the outcome, right? :-) It didn't mean it as a heuristic, but as something that (at least for many distributions) would be implemented explicitly.

**Comment by mschauer** [comment]
We know how to deal with that, we have exactly the same behaviour, the same issues for iterators and we introduced traits `Base.EltypeUnknown` accordingly but I agree that this is just rehashing the same discussion in a new place.

**Comment by cscherrer** [comment]
> Yes, as mentioned multiple times

Ok, I saw discussion *that* it was that way, but didn't see anything about *why* it was that way.

The details you give are helpful for this, and it looks like we just expect entirely different use cases for the function. I think of eltype as answering "what type of values does this container hold?" (distributions are a kind of container). In some cases, `Any` might be all we know.

It seems you expect it to be more like, "What primitive type should be used to instantiate arrays constructed using this?". That a fine question, just very different than I expected. From the many issues and PRs, it seems pretty common for people to be surprised by this usage.

**Comment by rfourquet** [comment]
As I alluded to in https://github.com/JuliaStats/Distributions.jl/pull/882#issuecomment-498994725, there is `Random.gentype` for the purpose described here. There was an issue about removing it (https://github.com/JuliaLang/julia/issues/31968), but I would like to close it because of a couple of use cases I have, and the situation in Distributions.jl suggests also to not remove it.

**Comment by mschauer** [comment]
> Now we have size returning the size of the variates and eltype returning the type of the dist params. :-)

One more of our original sins.

**Comment by oschulz** [comment]
> and the situation in Distributions.jl suggests also to not remove it.

So we'd define `gentype` for distributions to return a/the (default) variate type?

**Comment by cscherrer** [comment]
This looks great! And for non-distributional things it seems to act like `eltype`:
```julia
julia> Random.gentype([randn(3) for j in 1:4])
Vector{Float64} (alias for Array{Float64, 1})
```

So maybe we drop `sampletype` and use this? We should find a place to discuss what laws it's expected to follow, so we can avoid ending up in the `eltype` situation again.

**Comment by oschulz** [comment]
I'd love to have some official way to provide sample type information. I'm currently expanding the concept of `NamedTuple`-distributions and similar in ValueShapes.jl, and were samples aren't just scalars and arrays anymore this capability would be very useful.
